exonum/src/node/whitelist.rs

```diff
@@ -12,60 +12,80 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::collections::BTreeSet;
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::collections::BTreeMap;
 
 use crypto::PublicKey;
+use messages::Connect;
+use blockchain::ValidatorKeys;
+use helpers::fabric::NodePublicConfig;
+use node::ConnectInfo;
 
 // TODO: Don't reload whitelisted_peers if path the same. (ECR-172)
 
 /// `Whitelist` is special set to keep peers that can connect to us.
 #[derive(Debug, Clone, Serialize, Deserialize, Default)]
 pub struct Whitelist {
     whitelist_enabled: bool,
-    #[serde(default)]
-    whitelisted_peers: BTreeSet<PublicKey>,
 
     #[serde(default)]
-    #[serde(skip_serializing)]
-    #[serde(skip_deserializing)]
-    validators_list: BTreeSet<PublicKey>,
+    peers: BTreeMap<PublicKey, SocketAddr>,
 }
 
 impl Whitelist {
     /// Returns `true` if a peer with the given public key can connect.
     pub fn allow(&self, peer: &PublicKey) -> bool {
-        !self.whitelist_enabled || self.validators_list.contains(peer)
-            || self.whitelisted_peers.contains(peer)
+        !self.whitelist_enabled || self.peers.contains_key(peer)
     }
 
     /// Adds peer to the whitelist.
-    pub fn add(&mut self, peer: PublicKey) {
-        self.whitelisted_peers.insert(peer);
+    pub fn add(&mut self, peer: ConnectInfo) {
+        self.peers.insert(peer.public_key, peer.addr);
     }
 
-    /// Returns list of whitelisted peers.
-    pub fn collect_allowed(&self) -> Vec<&PublicKey> {
-        self.whitelisted_peers
+    /// Returns `true` if whitelist is enabled, otherwise everyone can connect.
+    pub fn is_enabled(&self) -> bool {
+        self.whitelist_enabled
+    }
+
+    /// Create whitelist from validators_keys and peers.
```

Comment:
It's obvious from signature. `Creates whitelist form validators keys and corresponding IP addresses`?